File: examples/spiral/spiral.js
let cols, rows, size;
let t = 0;
let background = "#000000";
let colors = ["#fcf6bd", "#d0f4de", "#a9def9", "#e4c1f9", "#ff99c8"];

function setup() {
  c.background(" ", background, background);
  c.cellPadding(" ");
}

function draw() {
  c.clear();
  size = 1;
  cols = c.width / size;
  rows = c.height / size;

  for (let i = 0; i < cols; i++) {
    for (let j = 0; j < cols; j++) {
      let x = i * size;
      let y = j * size;
      let d = c.distance(x, y, c.width / 2, c.height / 2);
      let k = 0.6;
      let dx = x - c.width / 2;
      let dy = y - c.height / 2;
      let angle = Math.atan2(dy, dx);
      let spiralPath = Math.sin(d / k + angle + t);
      let df = 2;
      let af = 2;
      threshold = Math.sin(d / df + af * angle);

      c.stroke("⬤", colorGradient(c.width, d), background);

      if (spiralPath > threshold) {
        c.point(x, y);
      }
    }
  }

  t += 0.5;
}

function colorGradient(width, d) {
  let step = width / colors.length;
  for (let i = 0; i < colors.length; i++) {
    if (d <= (i + 1) * step) {
      return colors[i];
    }
  }
  return colors[colors.length - 1];
}

File: examples/spiral/spiral.go
package main

import (
	"context"
	"math"

	"github.com/emprcl/runal"
)

var (
	t          float64
	background string   = "#000000"
	colors     []string = []string{
		"#fcf6bd",
		"#d0f4de",
		"#a9def9",
		"#e4c1f9",
		"#ff99c8",
	}
)

func main() {
	runal.Run(context.Background(), setup, draw)
}

func setup(c *runal.Canvas) {
	c.Background(" ", background, background)
	c.CellPadding(" ")
}

func draw(c *runal.Canvas) {
	c.Clear()
	size := 1
	cols := int(math.Round(float64(c.Width) / float64(size)))
	rows := int(math.Round(float64(c.Height) / float64(size)))
	for i := 0; i < cols; i++ {
		for j := 0; j < rows; j++ {
			x := i * size
			y := j * size
			d := c.Dist(x, y, c.Width/2, c.Height/2)
			k := .6
			dx := float64(x) - float64(c.Width)/2.
			dy := float64(y) - float64(c.Height)/2.
			angle := math.Atan2(dy, dx)
			spiralPath := math.Sin(d/k + angle + t)
			df := 2.
			af := 2.
			threshold := math.Sin(d/df + af*angle)

			c.Stroke("⬤", colorGradient(c.Width, d), background)

			if spiralPath > threshold {
				c.Point(x, y)
			}
		}
	}

	t += 0.5
}

func colorGradient(width int, d float64) string {
	step := width / len(colors)
	for i := 0; i < len(colors); i++ {
		if d <= float64((i+1)*step) {
			return colors[i]
		}
	}
	return colors[len(colors)-1]
}
